File: src/p0528/java/solution.java
import java.util.Random;

class Solution {

    private int[] sum;
    private int total;
    private Random random;

    public Solution(int[] w) {
        sum = w;
        total = sum[0];
        for (int i = 1; i < sum.length; i++) {
            total += sum[i];
            sum[i] += sum[i - 1];
        }
        random = new Random();
    }

    public int pickIndex() {
        int r = random.nextInt(total);
        int upper = 0, lower = sum.length - 1;
        while (upper <= lower) {
            int mid = (upper + lower) / 2;
            if (sum[mid] <= r) {
                upper = mid + 1;
            } else {
                lower = mid - 1;
            }
        }
        return upper;
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(w);
 * int param_1 = obj.pickIndex();
 */

File: src/p0528/cpp/solution.cpp
#include <iostream>
#include <vector>
#include <random>
#include <algorithm>

using namespace std;

class Solution {
public:
    Solution(vector<int> w) : sum(w) {
        total = sum[0];
        for (auto i = 1; i < sum.size(); i++) {
            total += sum[i];
            sum[i] += sum[i - 1];
        }
        mt = mt19937(rd());
        dist = uniform_int_distribution<>(0, total - 1);
    }

    int pickIndex() {
        return upper_bound(sum.begin(), sum.end(), dist(mt)) - sum.begin();
    }

private:
    vector<int> sum;
    int total;
    random_device rd;
    mt19937 mt;
    uniform_int_distribution<> dist;
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(w);
 * int param_1 = obj.pickIndex();
 */